File: example-kt-03adapters/src/androidTest/java/foo/bar/example/foreadapterskt/ui/playlist/PlaylistsViewTest.java
package foo.bar.example.foreadapterskt.ui.playlist;

import android.app.Activity;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import androidx.test.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnit4;
import foo.bar.example.foreadapterskt.ui.EspressoTestMatchers;
import foo.bar.example.foreadapterskt.R;
import foo.bar.example.foreadapterskt.feature.playlist.PlaylistAdvancedModel;
import foo.bar.example.foreadapterskt.feature.playlist.PlaylistSimpleModel;
import foo.bar.example.foreadapterskt.feature.playlist.Track;

import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_PORTRAIT;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isEnabled;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.not;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

/**
 * Here we make sure that the view elements accurately reflect the state of the models
 * and that clicking the buttons results in the correct action being performed
 */
@RunWith(AndroidJUnit4.class)
public class PlaylistsViewTest {


    private PlaylistSimpleModel mockPlaylistSimpleModel;
    private PlaylistAdvancedModel mockPlaylistAdvancedModel;


    @Before
    public void setup(){

        //MockitoAnnotations.initMocks(WalletView.this);
        System.setProperty("dexmaker.dexcache", InstrumentationRegistry.getTargetContext().getCacheDir().getPath());

        mockPlaylistAdvancedModel = mock(PlaylistAdvancedModel.class);
        mockPlaylistSimpleModel = mock(PlaylistSimpleModel.class);
    }


    @Test
    public void emptyPlaylists() throws Exception {

        //arrange
        new StateBuilder(mockPlaylistAdvancedModel, mockPlaylistSimpleModel)
                .withAdvancedPlaylistHavingTracks(0)
                .withSimplePlaylistHavingTracks(0)
                .withPlaylistsContainingTracks(new Track(R.color.pastel1))
                .createRule()
                .launchActivity(null);

        //act

        //assert
        onView(withId(R.id.playlist_add1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_add2_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_clear1_button)).check(matches(not(isEnabled())));
        onView(withId(R.id.playlist_clear2_button)).check(matches(not(isEnabled())));

        onView(withId(R.id.playlist_addMany1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_addMany2_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_removeMany1_button)).check(matches(not(isEnabled())));
        onView(withId(R.id.playlist_removeMany2_button)).check(matches(not(isEnabled())));

        onView(withId(R.id.playlist_totaltracks1_textview)).check(matches(withText("[0]")));
        onView(withId(R.id.playlist_totaltracks2_textview)).check(matches(withText("[0]")));

        onView(withId(R.id.playlist_list1_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(0)));
        onView(withId(R.id.playlist_list2_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(0)));

    }


    @Test
    public void with3ItemsInEachPlaylist() throws Exception {

        //arrange
        new StateBuilder(mockPlaylistAdvancedModel, mockPlaylistSimpleModel)
                .withAdvancedPlaylistHavingTracks(3)
                .withSimplePlaylistHavingTracks(3)
                .withPlaylistsContainingTracks(new Track(R.color.pastel1))
                .createRule()
                .launchActivity(null);

        //act

        //assert
        onView(withId(R.id.playlist_add1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_add2_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_clear1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_clear2_button)).check(matches(isEnabled()));

        onView(withId(R.id.playlist_addMany1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_addMany2_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_removeMany1_button)).check(matches(not(isEnabled())));
        onView(withId(R.id.playlist_removeMany2_button)).check(matches(not(isEnabled())));

        onView(withId(R.id.playlist_totaltracks1_textview)).check(matches(withText("[3]")));
        onView(withId(R.id.playlist_totaltracks2_textview)).check(matches(withText("[3]")));

        onView(withId(R.id.playlist_list1_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(3)));
        onView(withId(R.id.playlist_list2_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(3)));

    }


    @Test
    public void with5ItemsInEachPlaylist() throws Exception {

        //arrange
        new StateBuilder(mockPlaylistAdvancedModel, mockPlaylistSimpleModel)
                .withAdvancedPlaylistHavingTracks(5)
                .withSimplePlaylistHavingTracks(5)
                .withPlaylistsContainingTracks(new Track(R.color.pastel1))
                .createRule()
                .launchActivity(null);

        //act

        //assert
        onView(withId(R.id.playlist_add1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_add2_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_clear1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_clear2_button)).check(matches(isEnabled()));

        onView(withId(R.id.playlist_addMany1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_addMany2_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_removeMany1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_removeMany2_button)).check(matches(isEnabled()));

        onView(withId(R.id.playlist_totaltracks1_textview)).check(matches(withText("[5]")));
        onView(withId(R.id.playlist_totaltracks2_textview)).check(matches(withText("[5]")));

        onView(withId(R.id.playlist_list1_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(5)));
        onView(withId(R.id.playlist_list2_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(5)));

    }


    @Test
    public void withDifferentItemsInEachPlaylist() throws Exception {

        //arrange
        new StateBuilder(mockPlaylistAdvancedModel, mockPlaylistSimpleModel)
                .withSimplePlaylistHavingTracks(0)
                .withAdvancedPlaylistHavingTracks(5)
                .withPlaylistsContainingTracks(new Track(R.color.pastel1))
                .createRule()
                .launchActivity(null);

        //act

        //assert
        onView(withId(R.id.playlist_clear1_button)).check(matches(not(isEnabled())));
        onView(withId(R.id.playlist_clear2_button)).check(matches(isEnabled()));

        onView(withId(R.id.playlist_removeMany1_button)).check(matches(not(isEnabled())));
        onView(withId(R.id.playlist_removeMany2_button)).check(matches(isEnabled()));

        onView(withId(R.id.playlist_totaltracks1_textview)).check(matches(withText("[0]")));
        onView(withId(R.id.playlist_totaltracks2_textview)).check(matches(withText("[5]")));

        onView(withId(R.id.playlist_list1_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(0)));
        onView(withId(R.id.playlist_list2_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(5)));

    }


    @Test
    public void stateMaintainedAfterRotation() throws Exception {

        //arrange
        Activity activity = new StateBuilder(mockPlaylistAdvancedModel, mockPlaylistSimpleModel)
                .withAdvancedPlaylistHavingTracks(3)
                .withSimplePlaylistHavingTracks(3)
                .withPlaylistsContainingTracks(new Track(R.color.pastel1))
                .createRule()
                .launchActivity(null);
        activity.setRequestedOrientation(SCREEN_ORIENTATION_LANDSCAPE);

        onView(withId(R.id.playlist_addMany1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_removeMany1_button)).check(matches(not(isEnabled())));
        onView(withId(R.id.playlist_totaltracks1_textview)).check(matches(withText("[3]")));

        //act
        activity.setRequestedOrientation(SCREEN_ORIENTATION_PORTRAIT);

        //assert
        onView(withId(R.id.playlist_addMany1_button)).check(matches(isEnabled()));
        onView(withId(R.id.playlist_removeMany1_button)).check(matches(not(isEnabled())));
        onView(withId(R.id.playlist_totaltracks1_textview)).check(matches(withText("[3]")));
    }


    @Test
    public void clickAddTrackAdvancedCallsModel() throws Exception {
        //arrange
        new StateBuilder(mockPlaylistAdvancedModel, mockPlaylistSimpleModel)
                .withAdvancedPlaylistHavingTracks(0)
                .withSimplePlaylistHavingTracks(0)
                .withPlaylistsContainingTracks(new Track(R.color.pastel1))
                .createRule()
                .launchActivity(null);

        //act
        onView(withId(R.id.playlist_add2_button)).perform(click());

        //assert
        verify(mockPlaylistAdvancedModel).addNewTrack();
    }


    @Test
    public void clickAddTrackSimpleCallsModel() throws Exception {
        //arrange
        new StateBuilder(mockPlaylistAdvancedModel, mockPlaylistSimpleModel)
                .withAdvancedPlaylistHavingTracks(0)
                .withSimplePlaylistHavingTracks(0)
                .withPlaylistsContainingTracks(new Track(R.color.pastel1))
                .createRule()
                .launchActivity(null);

        //act
        onView(withId(R.id.playlist_add1_button)).perform(click());

        //assert
        verify(mockPlaylistSimpleModel).addNewTrack();
    }

}

File: example-kt-03adapters/src/androidTest/java/foo/bar/example/foreadapterskt/ui/playlist/PlaylistsViewTest.kt
package foo.bar.example.foreadapterskt.ui.playlist

import android.content.pm.ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
import android.content.pm.ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
import android.view.View
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.isEnabled
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.ext.junit.runners.AndroidJUnit4
import foo.bar.example.foreadapterskt.R
import foo.bar.example.foreadapterskt.feature.playlist.mutable.MutablePlaylistModel
import foo.bar.example.foreadapterskt.feature.playlist.immutable.ImmutablePlaylistModel
import foo.bar.example.foreadapterskt.feature.playlist.Track
import foo.bar.example.foreadapterskt.ui.EspressoTestMatchers
import io.mockk.MockKAnnotations
import io.mockk.impl.annotations.MockK
import io.mockk.verify
import org.hamcrest.Matchers.not
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Here we make sure that the view elements accurately reflect the state of the models
 * and that clicking the buttons results in the correct action being performed
 */
@RunWith(AndroidJUnit4::class)
class PlaylistsViewTest {


    @MockK
    private lateinit var mockImmutablePlaylistModel: ImmutablePlaylistModel
    @MockK
    private lateinit var mockMutablePlaylistModel: MutablePlaylistModel


    @Before
    fun setUp() = MockKAnnotations.init(this, relaxed = true)


    @Test
    @Throws(Exception::class)
    fun emptyPlaylists() {

        //arrange
        StateBuilder(mockMutablePlaylistModel, mockImmutablePlaylistModel)
            .withUpdatablePlaylistHavingTracks(0)
            .withDiffablePlaylistHavingTracks(0)
            .createRule()
            .launchActivity(null)

        //act

        //assert
        onView(withId(R.id.updatable_add_button)).check(matches(isEnabled()))
        onView(withId(R.id.updatable_add5_button)).check(matches(isEnabled()))
        onView(withId(R.id.updatable_remove5_button)).check(matches(not<View>(isEnabled())))
        onView(withId(R.id.updatable_clear_button)).check(matches(not<View>(isEnabled())))
        onView(withId(R.id.updatable_totaltracks_textview)).check(matches(withText("0")))
        
        onView(withId(R.id.diffable_add_button)).check(matches(isEnabled()))
        onView(withId(R.id.diffable_add5_button)).check(matches(isEnabled()))
        onView(withId(R.id.diffable_remove5_button)).check(matches(not<View>(isEnabled())))
        onView(withId(R.id.diffable_clear_button)).check(matches(not<View>(isEnabled())))
        onView(withId(R.id.diffable_totaltracks_textview)).check(matches(withText("0")))

        onView(withId(R.id.updatable_list_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(0)))
        onView(withId(R.id.diffable_list_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(0)))

    }


    @Test
    @Throws(Exception::class)
    fun with3ItemsInEachPlaylist() {

        //arrange
        StateBuilder(mockMutablePlaylistModel, mockImmutablePlaylistModel)
            .withUpdatablePlaylistHavingTracks(3)
            .withDiffablePlaylistHavingTracks(3)
            .withPlaylistsContainingTrack(Track(R.color.pastel1, 123))
            .createRule()
            .launchActivity(null)

        //act

        //assert
        onView(withId(R.id.updatable_add_button)).check(matches(isEnabled()))
        onView(withId(R.id.diffable_add_button)).check(matches(isEnabled()))
        onView(withId(R.id.updatable_clear_button)).check(matches(isEnabled()))
        onView(withId(R.id.diffable_clear_button)).check(matches(isEnabled()))

        onView(withId(R.id.updatable_add5_button)).check(matches(isEnabled()))
        onView(withId(R.id.diffable_add5_button)).check(matches(isEnabled()))
        onView(withId(R.id.updatable_remove5_button)).check(matches(not<View>(isEnabled())))
        onView(withId(R.id.diffable_remove5_button)).check(matches(not<View>(isEnabled())))

        onView(withId(R.id.updatable_totaltracks_textview)).check(matches(withText("3")))
        onView(withId(R.id.diffable_totaltracks_textview)).check(matches(withText("3")))

        onView(withId(R.id.updatable_list_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(3)))
        onView(withId(R.id.diffable_list_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(3)))

    }


    @Test
    @Throws(Exception::class)
    fun with5ItemsInEachPlaylist() {

        //arrange
        StateBuilder(mockMutablePlaylistModel, mockImmutablePlaylistModel)
            .withUpdatablePlaylistHavingTracks(5)
            .withDiffablePlaylistHavingTracks(5)
            .withPlaylistsContainingTrack(Track(R.color.pastel1, 123))
            .createRule()
            .launchActivity(null)

        //act

        //assert
        onView(withId(R.id.updatable_add_button)).check(matches(isEnabled()))
        onView(withId(R.id.diffable_add_button)).check(matches(isEnabled()))
        onView(withId(R.id.updatable_clear_button)).check(matches(isEnabled()))
        onView(withId(R.id.diffable_clear_button)).check(matches(isEnabled()))

        onView(withId(R.id.updatable_add5_button)).check(matches(isEnabled()))
        onView(withId(R.id.diffable_add5_button)).check(matches(isEnabled()))
        onView(withId(R.id.updatable_remove5_button)).check(matches(isEnabled()))
        onView(withId(R.id.diffable_remove5_button)).check(matches(isEnabled()))

        onView(withId(R.id.updatable_totaltracks_textview)).check(matches(withText("5")))
        onView(withId(R.id.diffable_totaltracks_textview)).check(matches(withText("5")))

        onView(withId(R.id.updatable_list_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(5)))
        onView(withId(R.id.diffable_list_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(5)))

    }


    @Test
    @Throws(Exception::class)
    fun withDifferentItemsInEachPlaylist() {

        //arrange
        StateBuilder(mockMutablePlaylistModel, mockImmutablePlaylistModel)
            .withDiffablePlaylistHavingTracks(0)
            .withUpdatablePlaylistHavingTracks(5)
            .withPlaylistsContainingTrack(Track(R.color.pastel1, 123))
            .createRule()
            .launchActivity(null)

        //act

        //assert
        onView(withId(R.id.diffable_clear_button)).check(matches(not<View>(isEnabled())))
        onView(withId(R.id.updatable_clear_button)).check(matches(isEnabled()))

        onView(withId(R.id.diffable_remove5_button)).check(matches(not<View>(isEnabled())))
        onView(withId(R.id.updatable_remove5_button)).check(matches(isEnabled()))

        onView(withId(R.id.diffable_totaltracks_textview)).check(matches(withText("0")))
        onView(withId(R.id.updatable_totaltracks_textview)).check(matches(withText("5")))

        onView(withId(R.id.diffable_list_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(0)))
        onView(withId(R.id.updatable_list_recycleview)).check(matches(EspressoTestMatchers.withRecyclerViewItems(5)))

    }


    @Test
    @Throws(Exception::class)
    fun stateMaintainedAfterRotation() {

        //arrange
        val activity = StateBuilder(mockMutablePlaylistModel, mockImmutablePlaylistModel)
            .withUpdatablePlaylistHavingTracks(3)
            .withDiffablePlaylistHavingTracks(3)
            .withPlaylistsContainingTrack(Track(R.color.pastel1, 123))
            .createRule()
            .launchActivity(null)
        activity.requestedOrientation = SCREEN_ORIENTATION_LANDSCAPE

        onView(withId(R.id.updatable_add5_button)).check(matches(isEnabled()))
        onView(withId(R.id.updatable_remove5_button)).check(matches(not<View>(isEnabled())))
        onView(withId(R.id.updatable_totaltracks_textview)).check(matches(withText("3")))

        //act
        activity.requestedOrientation = SCREEN_ORIENTATION_PORTRAIT

        //assert
        onView(withId(R.id.updatable_add5_button)).check(matches(isEnabled()))
        onView(withId(R.id.updatable_remove5_button)).check(matches(not<View>(isEnabled())))
        onView(withId(R.id.updatable_totaltracks_textview)).check(matches(withText("3")))
    }


    @Test
    @Throws(Exception::class)
    fun clickAddTrackDiffableCallsModel() {
        //arrange
        StateBuilder(mockMutablePlaylistModel, mockImmutablePlaylistModel)
            .withUpdatablePlaylistHavingTracks(0)
            .withDiffablePlaylistHavingTracks(0)
            .createRule()
            .launchActivity(null)

        //act
        onView(withId(R.id.diffable_add_button)).perform(click())

        //assert
        verify(exactly = 1) {
            mockImmutablePlaylistModel.addNTracks(1)
        }
    }


    @Test
    @Throws(Exception::class)
    fun clickAddTrackUpdatableCallsModel() {
        //arrange
        StateBuilder(mockMutablePlaylistModel, mockImmutablePlaylistModel)
            .withUpdatablePlaylistHavingTracks(0)
            .withDiffablePlaylistHavingTracks(0)
            .createRule()
            .launchActivity(null)

        //act
        onView(withId(R.id.updatable_add_button)).perform(click())

        //assert
        verify(exactly = 1) {
            mockMutablePlaylistModel.addNTracks(1)
        }
    }

}
